tslearn/early_classification.py

```diff
@@ -0,0 +1,266 @@
+        distances_clusters = np.linalg.norm(diffs, axis=(1, 2))
+        minimum_distance = np.min(distances_clusters)
+        average_distance = np.mean(distances_clusters)
+        minimum_distance_to_average = (
+            average_distance - minimum_distance
+        ) / average_distance
+        delta = (average_distance - distances_clusters) / average_distance
+        probas = 1.0 / (
+            np.exp(self.lamb * minimum_distance_to_average)
+            + np.exp(self.lamb * (minimum_distance_to_average - delta))
+            + 1e-6
+        )
+        return probas / probas.sum()
+
+    def _expected_cost(self, X, tau):
+        """
+        From a incomplete series xt, this function compute the expected cost of a prediction made at time "last time of
+        xt + tau"
+
+        Parameters
+        ----------
+        X: vector
+            a time series that may be truncated for which we want to have the cost
+        tau: int
+            gives the future time for which the cost is calculated
+
+        Returns
+        --------
+        float : the computed cost
+        """
+        proba_clusters = self.get_cluster_probas(X=X)
+        truncated_t = X.shape[-1]
+        sum_pyhatyck = np.sum(self.pyhatyck_[truncated_t + tau - self.minimum_time_stamp], axis=1)
+        sum_pyhatyck = np.transpose(sum_pyhatyck)
+        sum_global = np.sum((sum_pyhatyck * self.pyck_), axis=0)
+        cost = np.dot(proba_clusters, sum_global)
+        return cost + self._cost_time(truncated_t + tau)
+
+    def cost_function_minimizer(self, end_of_time, xt):
+        """
+        We want to minimize a function "funct" according a time series "xt" for a list of integers from 0 to
+        "end_of_time"
+
+        Parameters
+        ----------
+        end_of_time: int
+            the highest integer at which the function to be minimized is calculated
+        function_to_minimize: funct
+            the function of interest
+        xt: vector
+            the time series related to the function "funct"
+
+        Returns
+         ------
+         int : The integer in {0,...,stop} that minimizes "funct"
+         float : the so fat minimum cost
+        """
+        costs = [self._expected_cost(xt, tau) for tau in range(end_of_time + 1)]
+        tau_star = np.argmin(costs)
+        min_cost = costs[tau_star]
+        return tau_star, min_cost
+
+    def _predict_single_series(self, Xi):
+        """
+        This function classifies a single time series xt
+
+        Parameters
+        ----------
+        xt: vector
+            a time series that is probably incomplete but that nonetheless we want to classify
+        Returns
+        -------
+        int: the class which is predicted
+        int : the time of the prediction
+        float : the probability used for computing the cost
+        float : the loss when classifying
+        """
+        length_Xi = Xi.shape[0]
+        Xi = np.reshape(Xi, (1, length_Xi))
```

Comment:
You should not do that, because the data can be multi dimensional and then it won't work.
I think the only problem you might have is when you call `predict` and `predict_proba`, but then you should just pass `[Xi[:t]]` as input (a list made of a single series).